### DISCUSSION TITLE
How can pdfplumber process pdf on char level

### DISCUSSION THREAD

**Discussion by NewUserHa** [jsvine/pdfplumber#1017] [Ask for help with specific PDFs]
Here I have a PDF with non-Unicode custom font (issue 1), and it has the wrong bbox (like 2 times than real) that already overlaps with each character (issue 2).

Currently, I'm using OCR the characters from each font to get correct maps:

to translate them (issue 1):
* in pdfminer.six, I achieved via `high_level.extract_pages()` + walking each object recursive via `custom function` to match to replace to the correct ones.
(or can also use `PDFPageInterpreter` with a custom `pdfminer.converter.TextConverter` class with an override `render_char()` function)
* in pdfplumber, I didn't find the ability.
- is there an easier way in pdfplumber?

to resolve the overlaps for adding space between characters (issue 2):
* in pdfminer.six, I tried to divide the width and height of `LTChar` by 2 or 3 in that custom `render_char()`  with `self.cur_item._objs[-1]` to try to trigger `LAParams(word_margin=0.1, all_texts=1)`, but it didn't work. 
* in pdfplumber, `x_tolerance=-2` worked, but it added space between every character.
I'm not familiar with how the `LAParams` algorithm works, but in its code, it seems that it only adds space in `word_margin` level, and `char_margin` is only for group texts.
- since both didn't work. how can solve the space issue easily?

Thanks

**Comment by jsvine** [comment]
Hi @NewUserHa. I'm not sure I fully understand the questions here; it might help to provide the PDF, the code, and/or screenshots. But at least for the first part, you might try something like this:

```python
def adjust_bbox(char):
  [your custom logic here]
  return new_char

page = pdf.pages[0]
new_chars = list(map(adjust_bbox, page.chars))
text = pdfplumber.utils.extract_text(new_chars)
```

**Reply by NewUserHa** [reply]
Thanks for your reply.

`pdfplumber.utils.extract_text` looks great, but seems it's not in the document.
But I haven't tried it so not sure if it will work.

there's the pdf: [WS_T 483.8-2016.pdf](https://github.com/jsvine/pdfplumber/files/13319733/WS_T.483.8-2016.pdf)
only the last page is the one needs to be looked at,
the issues of the pdf are:
* used a custom font with a not unicode custom mapping
* faked bboxs for some characters of some fonts.

**Reply by jsvine** [reply]
Thanks for providing that PDF, which helps clarify the question. Let us know if the approach suggested above works for you.

And, correct, `pdfplumber.utils.extract_text` is not in the documentation. It can be useful, but it's not the main method I'd suggest people to use. I could see adding some more information about it in an "Advanced" section of the documentation.
